File: wordwiz/src/main/java/com/pyamsoft/wordwiz/WordWiz.java
/*
 * Copyright 2016 Peter Kenji Yamanaka
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.pyamsoft.wordwiz;

import android.app.Application;
import android.support.annotation.CheckResult;
import android.support.annotation.NonNull;
import android.support.v4.app.Fragment;
import com.android.annotations.Nullable;
import com.pyamsoft.pydroid.ui.PYDroidApplication;
import com.squareup.leakcanary.LeakCanary;
import com.squareup.leakcanary.RefWatcher;

public class WordWiz extends PYDroidApplication {

  @Nullable private RefWatcher refWatcher;

  @CheckResult @NonNull public static RefWatcher getRefWatcher(@NonNull Fragment fragment) {
    final Application application = fragment.getActivity().getApplication();
    if (application instanceof WordWiz) {
      return ((WordWiz) application).getWatcher();
    } else {
      throw new IllegalStateException("Application is not WordWiz");
    }
  }

  @Override protected boolean exitBeforeInitialization() {
    return LeakCanary.isInAnalyzerProcess(this);
  }

  @Override protected void onCreateInDebugMode() {
    super.onCreateInDebugMode();
    refWatcher = LeakCanary.install(this);
  }

  @Override protected void onCreateInReleaseMode() {
    super.onCreateInReleaseMode();
    refWatcher = RefWatcher.DISABLED;
  }

  @NonNull @CheckResult private RefWatcher getWatcher() {
    if (refWatcher == null) {
      throw new IllegalStateException("RefWatcher is NULL");
    }
    return refWatcher;
  }
}

File: wordwiz/src/main/java/com/pyamsoft/wordwiz/WordWiz.kt
/*
 * Copyright 2019 Peter Kenji Yamanaka
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

package com.pyamsoft.wordwiz

import android.app.Application
import com.pyamsoft.pydroid.ui.PYDroid
import com.pyamsoft.pydroid.ui.PYDroid.Instance
import com.squareup.leakcanary.LeakCanary
import com.squareup.leakcanary.RefWatcher

class WordWiz : Application(), Instance {

  private var pyDroid: PYDroid? = null
  private lateinit var refWatcher: RefWatcher
  private lateinit var component: WordWizComponent

  override fun onCreate() {
    super.onCreate()
    if (LeakCanary.isInAnalyzerProcess(this)) {
      return
    }

    if (BuildConfig.DEBUG) {
      refWatcher = LeakCanary.install(this)
    } else {
      refWatcher = RefWatcher.DISABLED
    }

    PYDroid.init(
        this,
        this,
        getString(R.string.app_name),
        "https://github.com/pyamsoft/wordwiz/issues",
        BuildConfig.VERSION_CODE,
        BuildConfig.DEBUG
    )
  }

  override fun getPydroid(): PYDroid? = pyDroid

  override fun setPydroid(instance: PYDroid) {
    pyDroid = instance.also {
      component = WordWizComponentImpl(this, it.modules())
    }
  }

  override fun getSystemService(name: String): Any {
    if (Injector.name == name) {
      return component
    } else {
      return super.getSystemService(name)
    }
  }

}
